internal/services/notification/runeventssender.go

```diff
@@ -0,0 +1,152 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/internal/util"
+	"agola.io/agola/services/runservice/types"
+)
+
+type runEventsSender struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started bool
+
+	log zerolog.Logger
+
+	t *testing.T
+
+	mu sync.Mutex
+
+	listenAddress string
+	exposedURL    string
+}
+
+func setupRunEventsSender(pctx context.Context, t *testing.T) *runEventsSender {
+	log := testutil.NewLogger(t)
+	ctx, cancel := context.WithCancel(pctx)
+
+	wr := &runEventsSender{ctx: ctx, t: t, log: log, cancel: cancel}
+
+	port, err := testutil.GetFreePort("localhost", true, false)
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	wr.listenAddress = net.JoinHostPort("localhost", port)
+	wr.exposedURL = fmt.Sprintf("http://%s", net.JoinHostPort("localhost", port))
+
+	if err := wr.start(); err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+
+		wr.stop()
+	}()
+
+	return wr
+}
+
+func (wr *runEventsSender) start() error {
+	wr.mu.Lock()
+	defer wr.mu.Unlock()
+
+	if wr.started {
+		return fmt.Errorf("runEventsSender already started")
+	}
+
+	router := mux.NewRouter()
+
+	router.Handle("/api/v1alpha/runs/events", newRunEventsHandler(wr.log)).Methods("GET")
+
+	httpServer := http.Server{
+		Addr:    wr.listenAddress,
+		Handler: router,
+	}
+
+	lerrCh := make(chan error)
+	go func() {
+		lerrCh <- httpServer.ListenAndServe()
+	}()
+
+	go func() {
+		select {
+		case <-wr.ctx.Done():
+			httpServer.Close()
+		case err := <-lerrCh:
+			if err != nil {
+				wr.log.Err(err).Send()
+			}
+		}
+	}()
+
+	wr.started = true
+
+	return nil
+}
+
+func (wr *runEventsSender) stop() {
+	wr.mu.Lock()
+	defer wr.mu.Unlock()
+
+	if !wr.started {
+		return
+	}
+
+	wr.cancel()
+	wr.started = false
+}
+
+type runEventsHandler struct {
+	log zerolog.Logger
+}
+
+func newRunEventsHandler(log zerolog.Logger) *runEventsHandler {
+	return &runEventsHandler{log: log}
+}
+
+func (h *runEventsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "text/event-stream")
+	w.Header().Set("Cache-Control", "no-cache")
+	w.Header().Set("Connection", "keep-alive")
+
+	runEvent := types.RunEvent{Sequence: 1, RunEventType: "testrunevents", DataVersion: webhookVersion}
```

Comment:
All of this code for just sending a single event. I'll make this more configurable (methods to add/set the events to send) and improve the tests to check more cases (more events etc...).